echopop/computation/biology.py

```diff
@@ -104,160 +128,751 @@ def sum_strata_weight( haul_data: pd.DataFrame ,
+        .agg( { 'biomass_sexed_aged_all': 'sum' ,
+                'biomass_sexed_aged_adult': 'sum' } )
+        .reset_index( )
+    )
+
+    ### Calculate biomass proportions across age-length
+    # ---- Sum weights within each length-age bin for each stratum
+    proportions_weight_length_age = (
+        proportions_weight_length_age_sex
+        .groupby( [ 'stratum_num' , 'species_id' , 'length_bin' , 'age_bin' ] )
+        .agg( { 'weight_all': 'sum' ,
+                'weight_adult': 'sum' } )
+        .reset_index( )
+    )
+
+    # ---- Sum stratum weights
+    # ---- All fish
+    proportions_weight_length_age[ 'total_weight_all' ] = (
+        proportions_weight_length_age.groupby( [ 'stratum_num' , 'species_id' ] )[ 'weight_all' ].transform( sum )
+    )
+
+    # ---- Adult fish
+    proportions_weight_length_age[ 'total_weight_adult' ] = (
+        proportions_weight_length_age.groupby( [ 'stratum_num' , 'species_id' ] )[ 'weight_adult' ].transform( sum )
+    )
+
+    # ---- Calculate proportions
+    # ---- All fish
+    proportions_weight_length_age[ 'proportion_weight_length_all' ] = (
+        proportions_weight_length_age.weight_all / proportions_weight_length_age.total_weight_all
+    ).fillna( 0 )    
+
+    # ---- Adult fish
+    proportions_weight_length_age[ 'proportion_weight_length_adult' ] = (
+        proportions_weight_length_age.weight_adult / proportions_weight_length_age.total_weight_adult
+    ).fillna( 0 )    
+
+    # ---- Index `aged_proportions`
+    index_aged_proportions = aged_proportions.filter( regex = "^(?=.*aged|stratum_num)(?!.*unaged)" ).set_index( 'stratum_num' )
+
+    # ---- Initialize dataframe
+    stratum_kriged_biomass = proportions_weight_length_age.set_index( 'stratum_num' )
+    stratum_kriged_biomass[ 'proportion_aged_weight_all' ] =  index_aged_proportions.proportion_aged_weight_all
+    stratum_kriged_biomass[ 'proportion_aged_weight_adult' ] = index_aged_proportions.proportion_aged_weight_adult
+    stratum_kriged_biomass[ 'biomass_kriged' ] = kriged_stratum_biomass
+
+    # ---- Apportioned biomass for all fish
+    stratum_kriged_biomass[ 'biomass_aged_all' ] = (  
+        stratum_kriged_biomass.biomass_kriged
+        * stratum_kriged_biomass.proportion_weight_length_all 
+        * stratum_kriged_biomass.proportion_aged_weight_all
+    ).fillna( 0 )
+
+    # ---- Apportioned biomass for adult fish
+    stratum_kriged_biomass[ 'biomass_aged_adult' ] = (  
+        stratum_kriged_biomass.biomass_kriged
+        * stratum_kriged_biomass.proportion_weight_length_adult
+        * stratum_kriged_biomass.proportion_aged_weight_adult
+    ).fillna( 0 )
+
+    # ---- Sum across strata to produce 'grand totals' for all fish
+    apportioned_total_kriged_biomass = (
+        stratum_kriged_biomass
+        .groupby( [ 'species_id' , 'length_bin' , 'age_bin' ] )
+        .agg( { 'biomass_aged_all': 'sum' ,
+                'biomass_aged_adult': 'sum' } )
+        .reset_index( )
+    )
```

Comment:
I am not sure why you need to go from L446 all the way here to calculate these quantities. Aren't these just sums over both sexes for each length-age bin, that you can compute directly from `apportioned_sexed_kriged_biomass`?

If there are special cases that prevents going from `apportioned_sexed_kriged_biomass` to `apportioned_total_kriged_biomass`, the test should reflect that.